src/main/scala/org/phenoscape/kb/Similarity.scala

```diff
@@ -242,17 +244,43 @@ object Similarity {
     }
   }
 
-  def pairwiseJaccardSimilarity(iris: Set[IRI]): Future[Seq[JaccardScore]] =
-    Future.sequence(iris.map(iri => classSubsumers(iri).map(iri -> _))).map { irisSubsumers =>
-      val irisToSubsumers = irisSubsumers.toMap
+  def pairwiseJaccardSimilarity(iris: Set[IRI],
+                                relations: Set[IRI],
+                                pathOpt: Option[Path]): Future[Seq[JaccardScore]] = {
+
+    val termSubsumerPairsFut = getTermSubsumerPairs(iris, relations, pathOpt)
+
+    val termSubsumersMapFut = for {
+      termSubsumerPairs <- termSubsumerPairsFut
+    } yield {
+      val groupedByTerm = termSubsumerPairs.groupBy(_._1)
+
+      groupedByTerm
+        .map { case (term, pairs) =>
+          val subsumers = pairs.map(_._2).toSet
+          Map(term -> subsumers)
+        }
+        .flatten
```

Comment:
Can you use `flatMap` instead of `map` and then `flatten`? Or just use `map` and don't create a `Map` inside for every entry?